Use new SerializationFeature.register API with GraalVM/Mandrel >=21.3

zakkak opened the pull request:
https://github.com/oracle/graal/pull/3050 introduces a number of changes
around the internal APIs that are being used by Quarkus to register
classes for serialization thus breaking serialization support when using
21.3-dev

These changes are:

1. `SerializationRegistry` class has been moved from
   `com.oracle.svm.core.jdk.serialize` to
   `com.oracle.svm.reflect.serialize`

2. `addReflections` method moved from
   `com.oracle.svm.reflect.serialize.hosted.SerializationFeature` to
   `com.oracle.svm.reflect.serialize.hosted.SerializationBuilder` and
   became `private`

https://github.com/oracle/graal/pull/3050 also introduces
`SerializationFeature.register(Class<?>... classes)` which is available
as a public API (not internal) and can save us from having to keep up
with the internal APIs as new releases come out. This PR leverages the
new API to register classes for serialization when using GraalVM/Mandrel >=21.3.

Closes https://github.com/quarkusio/quarkus/issues/19338

zakkak committed:
Use new SerializationFeature.register API with GraalVM/Mandrel >=21.3

https://github.com/oracle/graal/pull/3050 introduces a number of changes
around the internal APIs that are being used by Quarkus to register
classes for serialization thus breaking serialization support when using
21.3-dev

These changes are:

1. `SerializationRegistry` class has been moved from
   `com.oracle.svm.core.jdk.serialize` to
   `com.oracle.svm.reflect.serialize`

2. `addReflections` method moved from
   `com.oracle.svm.reflect.serialize.hosted.SerializationFeature` to
   `com.oracle.svm.reflect.serialize.hosted.SerializationBuilder` and
   became `private`

https://github.com/oracle/graal/pull/3050 also introduces
`SerializationFeature.register(Class<?>... classes)` which is available
as a public API (not internal) and can save us from having to keep up
with the internal APIs as new releases come out. This PR leverages the
new API to register classes for serialization when using GraalVM/Mandrel
>=21.3.

Closes https://github.com/quarkusio/quarkus/issues/19338

zakkak commented:
CI run with 21.3-dev: https://github.com/graalvm/mandrel/actions/runs/1148784045 (please don't merge before it completes)

zakkak commented:
The tests are looking good, but I am switching this to draft to make it use `SerializationFeature.register(Class<?>... classes)` with GraalVM >=21.3

galderz commented:
Great work with the fix and thanks for spotting this early on @zakkak! 👏

zakkak commented:
New CI run to test branch with 21.3: https://github.com/graalvm/mandrel/actions/runs/1150563667

gsmet commented:
Yeah, thanks @zakkak ! I'll have a look early next week.

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java:
```diff
@@ -618,26 +656,22 @@ private MethodDescriptor createRegisterSerializationForClassMethod(ClassCreator
-                ofMethod("com.oracle.svm.reflect.serialize.hosted.SerializationFeature", "addReflections", void.class,
-                        Class.class, Class.class),
-                clazz, objectClass);
+        ResultHandle addRedlectionsArgs2 = tc.newArray(Class.class, tc.load(2));
```
```suggestion
        ResultHandle addReflectionsArgs2 = tc.newArray(Class.class, tc.load(2));
```

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java:
```diff
@@ -618,26 +656,22 @@ private MethodDescriptor createRegisterSerializationForClassMethod(ClassCreator
+        ResultHandle addRedlectionsArgs2 = tc.newArray(Class.class, tc.load(2));
+        tc.writeArrayValue(addRedlectionsArgs2, 0, clazz);
+        tc.writeArrayValue(addRedlectionsArgs2, 1, objectClass);
+        tc.invokeVirtualMethod(invokeMethodDescriptor, addReflectionsLookupMethod, tc.loadNull(), addRedlectionsArgs2);
```
```suggestion
        tc.invokeVirtualMethod(invokeMethodDescriptor, addReflectionsLookupMethod, tc.loadNull(), addReflectionsArgs2);
```

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java:
```diff
@@ -618,26 +656,22 @@ private MethodDescriptor createRegisterSerializationForClassMethod(ClassCreator
+        ResultHandle addRedlectionsArgs2 = tc.newArray(Class.class, tc.load(2));
+        tc.writeArrayValue(addRedlectionsArgs2, 0, clazz);
+        tc.writeArrayValue(addRedlectionsArgs2, 1, objectClass);
+        tc.invokeVirtualMethod(invokeMethodDescriptor, addReflectionsLookupMethod, tc.loadNull(), addRedlectionsArgs2);
```
What is the `tc.loadNull()` supposed to be?

zakkak replied:
`invoke`'s first argument is the `this` object, so since this is a static method I am passing `null` as the first argument.

> If the underlying method is static, then the specified obj argument is ignored. It may be null. 
See https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/Method.html#invoke(java.lang.Object,java.lang.Object...)

gsmet reviewed (changes requested):
Spotted a small typo and added a question.

@zakkak I suppose we want this in 2.2?

zakkak commented:
> Spotted a small typo and added a question.

Good catch! Fixed and answered respectively :)

> @zakkak I suppose we want this in 2.2?

Yes we will eventually need it in 2.2.

gsmet commented:
Thanks, nice work!
